ambry-router/src/main/java/com.github.ambry.router/PutOperation.java

```diff
@@ -373,6 +391,57 @@ void fillChunks() {
   }
 
   /**
+   * Submits encrypt job for the given {@link PutChunk} and processes the callback for the same
+   * @param putChunk the {@link PutChunk} that needs to be encrypted
+   * @param encryptContent {@code true} if content needs to be encrypted. {@code false} otherwise
+   */
+  private void encryptChunk(PutChunk putChunk, boolean encryptContent) {
+    try {
+      logger.trace("Chunk at index {} state moves to {}", putChunk.chunkIndex, ChunkState.Encrypting);
+      putChunk.state = ChunkState.Encrypting;
+      if (encryptContent) {
+        putChunk.buf.flip();
+        putChunk.chunkBlobSize = putChunk.buf.remaining();
+      }
+      logger.trace("Submitting encrypt job for chunk at index " + putChunk.chunkIndex);
+      exec.submitJob(new EncryptJob(passedInBlobProperties.getAccountId(), passedInBlobProperties.getContainerId(),
+          encryptContent ? putChunk.buf : null, ByteBuffer.wrap(putChunk.chunkUserMetadata), kms.getRandomKey(),
+          cryptoService, kms, (EncryptJob.EncryptJobResult result, Exception exception) -> {
+        logger.trace("Processing encrypt job callback for chunk at index " + putChunk.chunkIndex);
+        if (exception == null && !isOperationComplete()) {
+          if (encryptContent) {
+            putChunk.buf = result.getEncryptedBlobContent();
+          }
+          putChunk.encryptedPerBlobKey = result.getEncryptedKey();
+          putChunk.chunkUserMetadata = result.getEncryptedUserMetadata().array();
+          logger.trace("Completing encrypt job result for chunk at index " + putChunk.chunkIndex);
+          putChunk.onFillComplete(true);
+          // todo: add metrics to track encryption time
```

Comment:
when is this going to happen? another patch?

Reply:
I have been Iteratively working on this patch. Just now completed all tests and so will work on this today. But yes, in this patch itself. but will wait until I get a good round of reviews for the core part and then will add metrics.